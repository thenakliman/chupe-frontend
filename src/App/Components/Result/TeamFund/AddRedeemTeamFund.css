.team-fund-add-redeem-container {
  display: flex;
  flex-direction: column;
  position: fixed;
  margin: 3% 20%;
  border: 1px solid #000;
  width: 30rem;
  height: 25rem;
}

.add-redeem-team-fund-select {
  display: flex;
  justify-content: space-evenly;
  padding: 1.5rem 5rem;
}

.add-redeem-input-amount {
  display: flex;
  justify-content: space-evenly;
  padding: 1.5rem 5rem;
}

.team-fund-amount-container {
  display: flex;
  justify-content: space-evenly;
  padding: 1.5rem 5rem;

}

.team-fund-submit-button {
  display: flex;
  justify-content: space-evenly;
  padding: 1.5rem 5rem;
}

.team-fund-amount {
  width: 10rem;
}

.team-fund-button {
  padding: 1rem 2rem;
  margin: 2rem 0;
  float: right;
  color: white;

  cursor: pointer;
}