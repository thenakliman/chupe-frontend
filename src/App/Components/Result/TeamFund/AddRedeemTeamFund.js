import React from 'react';
import propTypes from 'prop-types';

require('./AddRedeemTeamFund.css');

/** Add or Redeem a team member fund. */
export class AddRedeemTeamFund extends React.Component {
  /** Constructor for the component.
   * @param {object} props for the component
   */
  constructor(props) {
    super(props);
    this.state = {
      fundType: '',
      amount: 0,
    };

    this.handleAmountUpdate = this.handleAmountUpdate.bind(this);
    this.handleFundType = this.handleFundType.bind(this);
    this.handleClick = this.handleClick.bind(this);
  }

  /** Fetch team fund type on mount. */
  componentWillMount() {
    this.props.fetchFundTypes();
  }

  /** Handle amount update.
   * @param {number} amount to be added
   */
  handleAmountUpdate(amount) {
    this.setState({amount: amount});
  }

  /** Handle fund type update.
   * @param {string} typeId
   */
  handleFundType(typeId) {
    const fund = this.props.fundTypes.find((fund) => fund.id == typeId);
    const defaultAmount = fund && fund.defaultAmount || 0;
    this.setState({fundType: typeId, amount: defaultAmount});
  }
  /** Handle submit event of button. */
  handleClick() {
    if (this.state.amount === 0) {
      return;
    }

    this.props.addFund(this.state.fundType, this.state.amount);
  }

  /** Return Add/Redeem team fund type component.
   * @return {object} component for add/redeem
   */
  render() {
    return (
      <div className={'team-fund-popup-container'}>
      <div id='team-fund-add-redeem-container-id'
           className='team-fund-add-redeem-container'>
        <div className='add-redeem-team-fund-select'>
          <select id='team-fund-select-type-id'
                  onChange={(event) => this.handleFundType(event.target.value)}
                  value={this.state.value}
                  >
            <option value=''>Select Fund Type</option>
            {
              this.props.fundTypes.map((fund) => (
                <option value={fund.id}
                        key={fund.id}>
                            {fund.description}
                </option>))
            }
          </select>
        </div>
        <div className='team-fund-amount-container'>
          Amount:
          <input id='team-fund-amount-id'
                 className='team-fund-amount'
                 value={this.state.amount}
                 onChange={
                  (event) => this.handleAmountUpdate(event.target.value)}
                 />
        </div>
        <div className='team-fund-submit-button'>
          <button id='team-fund-cancel-button-id'
                  className='team-fund-button'
                  onClick={(event) => this.props.closePopup()}>
                Cancel
          </button>
          <button id='team-fund-submit-button-id'
                  className='team-fund-button'
                  onClick={() => this.handleClick()}>
                Submit
          </button>
        </div>
      </div>
    </div>
    );
  }
}

AddRedeemTeamFund.propTypes = {
  fetchFundTypes: propTypes.func.isRequired,
  fundTypes: propTypes.arrayOf(propTypes.object).isRequired,
  addFund: propTypes.func.isRequired,
  closePopup: propTypes.func.isRequired,
};
