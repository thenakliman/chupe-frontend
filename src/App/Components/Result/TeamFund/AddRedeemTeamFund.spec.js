/* eslint-disable */
import React from 'react';
import {AddRedeemTeamFund} from './AddRedeemTeamFund';
/* eslint-enable */
import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';

describe('Team fund component snapshot', () => {
    it('should match the snapshot', () => {
        const wrapper = shallow(<AddRedeemTeamFund fetchFundTypes={()=>{}}
                                                   addFund={jest.fn()}
                                                   closePopup={()=>{}}
                                                   fundTypes={[]}/>);
        expect(toJson(wrapper)).toMatchSnapshot();
    });
});

describe('Add Redeem Team Fund component', () => {
    const fundTypes = [{type: 1, id: 1}, {type: 't2', id: 2}];
    it('Should have container div', () => {
        const wrapper = shallow(<AddRedeemTeamFund fetchFundTypes={()=>{}}
                                                   addFund={jest.fn()}
                                                   closePopup={()=>{}}
                                                   fundTypes={fundTypes}/>);

        expect(wrapper.find('#team-fund-add-redeem-container-id').length
            ).toEqual(1);
    });

    it('Should fetch team fund type on component will mount', () => {
        const fetchFundTypes = jest.fn();

        shallow(<AddRedeemTeamFund fetchFundTypes={fetchFundTypes}
                                   closePopup={()=>{}}
                                   addFund={jest.fn()}
                                   fundTypes={fundTypes}/>);

        expect(fetchFundTypes).toHaveBeenCalledWith();
    });

    it('Should have select field for add or redeem', () => {
        const wrapper = shallow(<AddRedeemTeamFund fetchFundTypes={()=>{}}
                                                   closePopup={()=>{}}
                                                   addFund={jest.fn()}
                                                   fundTypes={fundTypes}/>);

        expect(wrapper.find('#team-fund-action-select-id').length).toEqual(1);
    });

    it('Should have two options for select field', () => {
        const wrapper = shallow(<AddRedeemTeamFund fetchFundTypes={()=>{}}
                                                   closePopup={()=>{}}
                                                   addFund={jest.fn()}
                                                   fundTypes={fundTypes}/>);

        expect(wrapper.find('#team-fund-action-select-id').children().length
            ).toEqual(3);
    });

    it('Should have select field type of fund', () => {
        const wrapper = shallow(<AddRedeemTeamFund fetchFundTypes={()=>{}}
                                                   closePopup={()=>{}}
                                                   addFund={jest.fn()}
                                                   fundTypes={fundTypes}/>);

        expect(wrapper.find('#team-fund-select-type-id').length).toEqual(1);
    });

    it('Should have select field options', () => {
        const wrapper = shallow(<AddRedeemTeamFund fetchFundTypes={()=>{}}
                                                   closePopup={()=>{}}
                                                   addFund={jest.fn()}
                                                   fundTypes={fundTypes}/>);

        expect(wrapper.find('#team-fund-select-type-id').children().length
            ).toEqual(3);
    });

    it('Should have input field for amount', () => {
        const wrapper = shallow(<AddRedeemTeamFund fetchFundTypes={()=>{}}
                                                   closePopup={()=>{}}
                                                   addFund={jest.fn()}
                                                   fundTypes={fundTypes}/>);

        expect(wrapper.find('#team-fund-amount-id').length).toEqual(1);
    });

    it('Should have Submit button', () => {
        const wrapper = shallow(<AddRedeemTeamFund fetchFundTypes={()=>{}}
                                                   closePopup={()=>{}}
                                                   addFund={jest.fn()}
                                                   fundTypes={fundTypes}/>);

        expect(wrapper.find('#team-fund-submit-button-id').length).toEqual(1);
    });

    it('Should have cancel button', () => {
        const wrapper = shallow(<AddRedeemTeamFund fetchFundTypes={()=>{}}
                                                   closePopup={()=>{}}
                                                   addFund={jest.fn()}
                                                   fundTypes={fundTypes}/>);

        expect(wrapper.find('#team-fund-cancel-button-id').length).toEqual(1);
    });

    it('should have transaction type set to empty', () => {
        const wrapper = shallow(<AddRedeemTeamFund fetchFundTypes={()=>{}}
                                                   closePopup={()=>{}}
                                                   addFund={jest.fn()}
                                                   fundTypes={fundTypes}/>);

        expect(wrapper.state().transactionType).toEqual('');
    });

    it('should have fund type set to empty', () => {
        const wrapper = shallow(<AddRedeemTeamFund fetchFundTypes={()=>{}}
                                                   closePopup={()=>{}}
                                                   addFund={jest.fn()}
                                                   fundTypes={fundTypes}/>);

        expect(wrapper.state().fundType).toEqual('');
    });

    it('should set amount to default value of transaction type', () => {
        const wrapper = shallow(<AddRedeemTeamFund fetchFundTypes={()=>{}}
                                                   closePopup={()=>{}}
                                                   addFund={jest.fn()}
                                                   fundTypes={fundTypes}/>);

        expect(wrapper.state().amount).toEqual(0);
    });

    it('should update amount in state on change of input', () => {
        const wrapper = shallow(<AddRedeemTeamFund fetchFundTypes={()=>{}}
                                                   closePopup={()=>{}}
                                                   addFund={jest.fn()}
                                                   fundTypes={fundTypes}/>);
        const amount = 1000;
        wrapper.find('#team-fund-amount-id').simulate(
          'change', {target: {value: amount}});

        expect(wrapper.state().amount).toEqual(amount);
    });

    it('should update transaction type on select', () => {
        const wrapper = shallow(<AddRedeemTeamFund fetchFundTypes={()=>{}}
                                                   closePopup={()=>{}}
                                                   addFund={jest.fn()}
                                                   fundTypes={fundTypes}/>);

        wrapper.find('#team-fund-action-select-id').simulate(
          'change', {target: {value: 'CREDIT'}});

        expect(wrapper.state().transactionType).toEqual('CREDIT');
    });

    it('should update fund type on select', () => {
        const wrapper = shallow(<AddRedeemTeamFund fetchFundTypes={()=>{}}
                                                   closePopup={()=>{}}
                                                   addFund={jest.fn()}
                                                   fundTypes={fundTypes}/>);

        wrapper.find('#team-fund-select-type-id').simulate(
            'change', {target: {value: 1}});

        expect(wrapper.state().fundType).toEqual(1);
    });

    it('should not submit on button click when amount is zero', () => {
        const addFund = jest.fn();
        const wrapper = shallow(<AddRedeemTeamFund fetchFundTypes={()=>{}}
                                                   closePopup={()=>{}}
                                                   addFund={addFund}
                                                   fundTypes={fundTypes}/>);

        wrapper.setState(
          {
            transactionType: 'CREDIT',
            fundType: 1,
            amount: 0,
          });

        wrapper.find('#team-fund-submit-button-id').simulate('click');
        expect(addFund).not.toHaveBeenCalled();
    });

    it('should submit on button click when amount is not zero', () => {
        const addFund = jest.fn();
        const fundType = 1;
        const transactionType = 'CREDIT';
        const amount = 107;
        const wrapper = shallow(<AddRedeemTeamFund fetchFundTypes={()=>{}}
                                                   addFund={addFund}
                                                   closePopup={()=>{}}
                                                   fundTypes={fundTypes}/>);

        wrapper.setState(
          {
              transactionType: transactionType,
              fundType: fundType,
              amount: amount,
          });

        wrapper.find('#team-fund-submit-button-id').simulate('click');
        expect(addFund).toHaveBeenCalledWith(transactionType, fundType, amount);
    });

    it('should close pop up on click of cancel button', () => {
        const closePopup = jest.fn();
        const wrapper = shallow(<AddRedeemTeamFund fetchFundTypes={()=>{}}
                                                   addFund={()=>{}}
                                                   closePopup={closePopup}
                                                   fundTypes={fundTypes}/>);

        wrapper.find('#team-fund-cancel-button-id').simulate('click');
        expect(closePopup).toHaveBeenCalledWith();
    });
});
