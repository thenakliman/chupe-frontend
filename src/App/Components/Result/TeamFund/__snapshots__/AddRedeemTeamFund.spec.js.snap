// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Team fund component snapshot should match the snapshot 1`] = `
<span
  className="team-fund-add-redeem-container"
  id="team-fund-add-redeem-container-id"
>
  <span
    className="add-redeem-team-fund-select"
  >
    <select
      id="team-fund-select-type-id"
      onChange={[Function]}
    >
      <option
        value=""
      >
        Select Fund Type
      </option>
    </select>
  </span>
  <span
    className="team-fund-amount-container"
  >
    Amount:
    <input
      className="team-fund-amount"
      id="team-fund-amount-id"
      onChange={[Function]}
      value={0}
    />
  </span>
  <span
    className="team-fund-submit-button"
  >
    <button
      className="team-fund-button"
      id="team-fund-cancel-button-id"
      onClick={[Function]}
    >
      Cancel
    </button>
    <button
      className="team-fund-button"
      id="team-fund-submit-button-id"
      onClick={[Function]}
    >
      Submit
    </button>
  </span>
</span>
`;
